[#260] Show failed nix command

heitor-lassarote opened the pull request:
This PR displays the command that failed when an error is produced after a command exits with a non-zero code. The output is also cleared a bit. Example:

```
🚀 ℹ️ [deploy] [INFO] Running checks for flake in examples/simple/flake.nix
🚀 ❌ [deploy] [ERROR] Failed to check deployment: Nix checking command resulted in a bad exit code: Some(1). The failed command is provided below:
Command { std: "nix" "flake" "check" "examples/simple/flake.nix", kill_on_drop: false }
The stderr output is provided below:
error: in flake URL 'examples/simple/flake.nix', 'flake.nix' is not a commit hash
```

heitor-lassarote committed:
[#260] Show failed Nix command on exit

Problem: Many errors result from attempts to build commands which might
error with a non-zero exit code. The failed commands are not displayed
to the user, however.

Solution: For every `([A-Za-z])Exit(Option<i32>)` error, change it to
`\1Exit(Option<i32>, String)`, where the `String` contains the failed
command. Format each `tokyo` command using `{:?}` and save it to the
corresponding error.

Note that `Command` cannot be copied or cloned, and it's not possible to
access a command's `std` field to get the underlying command, so we
resort to putting the `Debug`-formatted command.

heitor-lassarote committed:
[#260] Abstract some of the boilerplate

Problem: It's common to spawn a command and occasionally fail during
running the command or because it exited, leading to code repetition.

Solution: Create a `command` module to abstract some of the boilerplate
by wrapping `tokio::process::Command`. Create a `run` function that
abstracts most of this boilerplate.

In some parts, the way commands are executed are a bit different, so
instead we leave them as it is.

heitor-lassarote committed:
[#260] Improve the output for failed Nix commands

Problem: The output when a command failed for deployment would show the
error message in the console output and _then_ the error message, which
was confusing.

Solution: Show the error message together with the error in an order
that makes sense. This is done by capturing the output in the error
message and printinting it in an exit error. We also use `Stdio::null`
rather than `Stdio::piped` to avoid printing to the console.

rvem commented on src/command.rs:
```diff
@@ -0,0 +1,131 @@
+        }
+    }
+
+    pub fn arg<S: AsRef<OsStr>>(&mut self, arg: S) -> &mut Command {
```
What is the motivation behind the reimplementation of `arg` and other methods defined for `TokioCommand` rather than creating a single `new` constructor that accepts already constructed `TokioCommand` and `run` to run the command handling the result?

rvem commented on src/bin/activate.rs:
```diff
@@ -362,17 +376,31 @@ pub async fn wait(temp_path: PathBuf, closure: String, activation_timeout: Optio
+}
+
 #[derive(Error, Debug)]
 pub enum ActivateError {
```
I like the fact that this error type became simpler

rvem commented on src/deploy.rs:
```diff
@@ -409,23 +437,35 @@ pub async fn deploy_profile(
 
         let ssh_activate_exit_status = ssh_activate_child
-            .wait()
+            .wait_with_output()
```
Is there a reason why the `run` method of `command::Command` is not used here instead of `wait_with_output`?

heitor-lassarote replied:
Because `ssh_activate_child` is a `tokio::process::Child` and not a `command::Command` or `tokio::process::Command`. Although there are patterns, the way the codebase handles children is not repetitive enough to easily abstract them into a `run` command for a new `command::Child`, but I can put some more thought into this if you'd like.

heitor-lassarote committed:
[#260] Fix bug with run deploy error formatting

Problem: Run deploy error should probably show the node name and the
error... but it duplicates the name in the place of the error.

Solution: Simply replace 0 with 1 in the formatting to use the error.

Sereja313 commented on src/command.rs:
```diff
@@ -0,0 +1,72 @@
+                };
+                write!(
+                    f,
+                    "{} command resulted in a bad exit code: {:?}. The failed command is provided below:\n{}\nThe stderr output is provided below:\n{}",
```
When nix source code is embedded into a command, it also embeds the indentation:
```
🚀 ❌ [deploy] [ERROR] Failed to evaluate deployment data: Nix eval command resulted in a bad exit code: Some(1). The failed command is provided below:
Command { std: "nix" "eval" "--json" ".#deploy" "--apply" "\n                      deploy:\n                      (deploy // {\n                        nodes = {\n                          inherit (deploy.nodes) \"skat\";\n                        };\n                      })\n                    ", kill_on_drop: false }
```
Perhaps we could use a heuristic algorithm that replaces a \n character followed by space characters with a single space. So it would like smth like:
```
Command { std: "nix" "eval" "--json" ".#deploy" "--apply" " deploy: (deploy // { nodes = { inherit (deploy.nodes) \"skat\"; }; }) ", kill_on_drop: false }
```

OTOH, this is debugging information, so consistency with the original may be more important than brevity, but since nix is ​​not sensitive to indentation this heuristic shouldn't cause any problems.

What do you think?

heitor-lassarote replied:
By the way, another alternative is to unescape characters, so `\n` for example turns into an actual newline and indentations keep making sense. What would you prefer?

heitor-lassarote replied:
Pushed a new commit improving the formatting.

heitor-lassarote replied:
Oops, committed a test example for accident, I deleted it in a fixup commit.

heitor-lassarote committed:
[#260] Format Nix commands

Problem: The current Nix command output from errors is difficult to
read. It's full of unescaped strings that could be made more
human-readable.

Solution: Use the `unespcape` crate to unescape the command, thus
pretty-printing it.
